import "building.proto";
import "position.proto";
import "role.proto";
import "warrior.proto";

package protocol;

// 战斗实体类型
enum CombatEntityType {
  COMBAT_ENTITY_TYPE_BUILDING = 1;
  COMBAT_ACTION_TYPE_WARRIOR = 2;
}

// 战斗目标
message CombatTarget {
  required CombatEntityType type = 1;                     // 类型
  required int32 id = 2;                                  // id
}

// 战斗动作类型
enum CombatActionType {
  COMBAT_ACTION_TYPE_BUILD = 1;                           // 建造
  COMBAT_ACTION_TYPE_MOVE = 2;                            // 移动
  COMBAT_ACTION_TYPE_LOCK_TARGET = 3;                     // 锁定目标
  COMBAT_ACTION_TYPE_DEATH = 4;                           // 死亡
  COMBAT_ACTION_TYPE_CONVERT_CAMP = 5;                    // 转换阵营
}

// 建造指令
// COMBAT_ACTION_TYPE_BUILD
message CombatBuildAction {
  required int32 building_id = 1;                         // 建筑id
  required int32 warrior_conf_id = 2;                     // 士兵配置id
  optional .protocol.CombatWarriorFields fields = 3;      // 士兵数据
}

// 移动指令
// COMBAT_ACTION_TYPE_MOVE
message CombatMoveAction {
  repeated int32 warrior_id = 1;                          // 士兵id集
  required .protocol.Position pos = 2;                    // 位置
  repeated .protocol.Position paths = 3;                  // 路径点
}

// 锁定目标指令
// COMBAT_ACTION_TYPE_LOCK_TARGET
message CombatLockTargetAction {
  required int32 warrior_id = 1;                          // 士兵id
  required int32 target_id = 2;                           // 目标id
}

// 死亡指令
// COMBAT_ACTION_TYPE_DEATH
message CombatDeathAction {
  required int32 warrior_id = 1;                          // 士兵id
}

// 转换阵营指令
// COMBAT_ACTION_TYPE_CONVERT_CAMP
message CombatConvertCampAction {
  required int32 id = 1;                                  // 建筑/士兵id
  required int32 camp_id = 2;                             // 阵营id
}

// 行动指令
message CombatAction {
  required int32 type = 1;                                // 操作类型(CombatActionType)
  required int32 timestamp = 2;                           // 时间戳
  optional CombatBuildAction build_action = 3;            // 建造指令
  optional CombatMoveAction move_action = 4;              // 移动指令
  optional CombatLockTargetAction lock_target_action = 5; // 锁定目标指令
  optional CombatDeathAction death_action = 6;            // 死亡指令
  optional CombatConvertCampAction convert_camp = 7;      // 转换阵营指令
}

// 行动序列
message CombatActionSequence {
  repeated CombatAction action = 1;                       // 操作序列
}

// 战斗状态镜像数据
message CombatStatusImage {
  repeated .protocol.CombatBuildingFields building = 1;   // 建筑
  repeated .protocol.CombatWarriorFields warrior = 2;     // 士兵
  repeated .protocol.CombatRoleFields role = 3;           // 角色
}
